Code/GraphMol/FileParsers/PNGParser.cpp

```diff
@@ -0,0 +1,293 @@
+#include <vector>
+#include <boost/crc.hpp>
+#include "FileParsers.h"
+#ifdef RDK_USE_BOOST_IOSTREAMS
+#include <zlib.h>
+#endif
+
+namespace RDKit {
+
+namespace PNGData {
+const std::string smilesTag = "rdkitSMILES";
+const std::string molTag = "rdkitMOL";
+const std::string pklTag = "rdkitPKL";
+}  // namespace PNGData
+
+namespace {
+std::vector<unsigned char> pngHeader = {137, 80, 78, 71, 13, 10, 26, 10};
+bool checkPNGHeader(std::istream &inStream) {
+  for (auto byte : pngHeader) {
+    unsigned char ibyte;
+    inStream.read((char *)&ibyte, 1);
+    if (ibyte != byte) {
+      return false;
+    }
+  }
+  return true;
+}
+}  // namespace
+
+std::map<std::string, std::string> PNGStreamToMetadata(std::istream &inStream) {
+  // confirm that it's a PNG file:
+  if (!checkPNGHeader(inStream)) {
+    throw FileParseException("PNG header not recognized");
+  }
+  std::map<std::string, std::string> res;
+  // the file is organized in chunks. Read through them until we find the tEXt
+  // block FIX: at some point we'll want to also include zEXt here, but that
+  // requires zlib
+  while (inStream) {
+    std::uint32_t blockLen;
+    inStream.read((char *)&blockLen, sizeof(blockLen));
+    if (inStream.fail()) {
+      throw FileParseException("error when reading from PNG");
+    }
+    // PNG is big endian, make sure we handle the order correctly
+    blockLen = EndianSwapBytes<BIG_ENDIAN_ORDER, HOST_ENDIAN_ORDER>(blockLen);
+    char bytes[4];
+    inStream.read(bytes, 4);
+    if (inStream.fail()) {
+      throw FileParseException("error when reading from PNG");
+    }
+    auto beginBlock = inStream.tellg();
+    if (bytes[0] == 'I' && bytes[1] == 'E' && bytes[2] == 'N' &&
+        bytes[3] == 'D') {
+      break;
+    }
+#ifndef RDK_USE_BOOST_IOSTREAMS
+    bool alreadyWarned = false;
+#endif
+    if (blockLen > 0 &&
+        ((bytes[0] == 't' && bytes[1] == 'E') ||
+         (bytes[0] == 'z' && bytes[1] == 'T')) &&
+        bytes[2] == 'X' && bytes[3] == 't') {
+      // in a tEXt block, read the key:
+      std::string key;
+      std::getline(inStream, key, '\0');
+      if (inStream.fail()) {
+        throw FileParseException("error when reading from PNG");
+      }
+      auto dataLen = blockLen - key.size() - 1;
+      std::string value;
+      if (bytes[0] == 't') {
+        value.resize(dataLen);
+        inStream.read(&value.front(), dataLen);
+        if (inStream.fail()) {
+          throw FileParseException("error when reading from PNG");
+        }
+      } else if (bytes[0] == 'z') {
+        value = "";
+#ifdef RDK_USE_BOOST_IOSTREAMS
+        std::unique_ptr<Bytef[]> compressedData(new Bytef[dataLen]);
+        inStream.read((char *)compressedData.get(), dataLen);
+        if (inStream.fail()) {
+          throw FileParseException("error when reading from PNG");
+        }
+        // we apparently have to guess how large the output data will be
+        uLongf uncompressedDataLen = dataLen * 10;
```

Comment:
I think you can use this idiom to get around the issue of not knowing the uncompressed length.

```
array_source src{v.data(), v.size()};
  filtering_istream is;
  is.push(zlib_decompressor{});
  is.push(src);
  std::string s;
  is >> s;
```